### PR TITLE
Assay Transform Script UX update to allow file drop and file path entry

### LINKED ISSUE #4952: Assay: Transform script UX update - implementation plan and dev
[Spec](https://docs.google.com/document/d/1V6BeH_vu5pWyTedrtRK11qLV9Ks5-c5_wmkY7cUeGvk/edit?usp=sharing)
[Github epic](https://github.com/LabKey/platform/issues/4823)

### Planning Tasks
- [x] functional review - 1hr @cnathe 
- [x] update spec - 2hr @wendyinnis 📍
- [x] design adjustments - 1hr @labkey-keegang  📍
- [x] implementation plan - 2h @cnathe 
- [x] dev review - 0.5h @labkey-klum 
- [x] test review - 0.5h @labkey-danield 

### Dev Tasks
- [x] add new container level `@scripts` dir with developer perm check - 5h
- [x] field editor add script UI update for drop file component - 3h
- [x] field editor add script UI update for “add via path” - 1h
- [x] field editor to validate script engine exists on add - 2h
- [x] transform script display via cards with name and path - 3h
- [x] transform script card remove option - 2h
- [x] filter out `@scripts` dir from file management page - 3h
- [x] new `@scripts` management page with files content component - 3h
- [x] add link to manage script page from field editor - 1h
- [x] add metrics - 1h
- [x] Update Spec - 1h 
- [x] Feature verification - 0.5h @bernarddlee 
- [x] Code Review - 1h @labkey-susanh 
- [x] Create manual test plan - 0.5h
- [x] Create automated test plan - 0.5h
- [x] Unit tests - 3h @cnathe 
- [x] Integration Tests - 3h @cnathe 
- [x] Manual testing - 3h @labkey-klum
- [x] Docs handoff - 0.5h
- [x] Metrics handoff  - 0.5h
- [x] Test review and merge - 1h

### PR DESCRIPTION
**Pull request by cnathe**
#### Rationale
Traditionally, adding a new transform script to an assay design requires placing the transform script in a location accessible to the server and manually adding the file path to the assay design. This PR adds a new option to the field editor so that a platform developer can drop a file onto the field editor UI and have that added to a new @scripts secure directory for a given LK container. 

#### Related Pull Requests
- https://github.com/LabKey/labkey-ui-components/pull/1351

#### Changes
- update @labkey/components package version
- More consistent use of constants in FileContentService 
  - CloudStoreService.CLOUD_NAME -> FileContentService.CLOUD_LINK
  - PipelineWebdavProvider.PIPELINE_LINK -> FileContentService.PIPELINE_LINK
  - "@files" -> FileContentService.FILES_LINK
- ScriptsResourceProvider for `@scripts` dir with perm check for user.isPlatformDeveloper
- Add new endpoint to find out if a script engine is configured for a given file extension

### CODE REVIEW

File: experiment/src/org/labkey/experiment/ScriptsResourceProvider.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+            return null;
+
+        Container c = folder.getContainer();
+        FileContentService service = FileContentService.get();
+        if (null == service)
+            return null;
+        if (service.isCloudRoot(c))
+            return null;
+
+        if (FileContentService.SCRIPTS_LINK.equalsIgnoreCase(name))
```
Comment (labkey-matthewb) [inline experiment/src/org/labkey/experiment/ScriptsResourceProvider.java]: redundant?

File: experiment/src/org/labkey/experiment/ScriptsResourceProvider.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+            {
+                AttachmentDirectory dir = service.getMappedAttachmentDirectory(c, FileContentService.ContentType.scripts, false);
+                if (dir != null)
+                {
+                    return new ScriptsResource(parent, name, dir.getFileSystemDirectory(), c.getPolicy());
+                }
+            }
+            catch (MissingRootDirectoryException e)
+            {
+                // Don't complain here, just hide the @files subfolder
```
Comment (labkey-matthewb) [inline experiment/src/org/labkey/experiment/ScriptsResourceProvider.java]:
```suggestion
                // Don't complain here, just hide the @scripts subfolder
```

**Review by labkey-matthewb — COMMENTED** [review]
The webdav parts look good to me.

File: experiment/src/org/labkey/experiment/ScriptsResourceProvider.java (hunk 1 above)
Comment (cnathe) [reply experiment/src/org/labkey/experiment/ScriptsResourceProvider.java]: yes, removed

**Comment by cnathe** [comment]
> The webdav parts look good to me.

Thanks for taking a look @labkey-matthewb. One thing I did notice is that if I go to the standard webdav viewer page as a platform dev, the @scripts directories are created as I start clicking around in the folder structure. Is there some way to prevent the creation of those extra / empty directories from this provider?

**Review by labkey-klum — APPROVED** [review]

**Comment by labkey-matthewb** [comment]
I just thought of one more thing.  We should make these act how we expect in search.  Probably we should exclude them?

**Comment by cnathe** [comment]
> We should make these act how we expect in search. Probably we should exclude them?

@labkey-matthewb we already have the following for the `ScriptsResourceProvider.ScriptsResource`. That should exclude them from search right?
```
        @Override
        public boolean shouldIndex()
        {
            return false;
        }
```

**Pull request merged**